sql/distsql: TestWindowFunctionsAgainstProcessor failed

sql/distsql.TestWindowFunctionsAgainstProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3216593&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3216593&tab=artifacts#/) on master @ [42ce12799466a06d6b41bfecc9ebc9ef367ab9c9](https://github.com/cockroachdb/cockroach/commits/42ce12799466a06d6b41bfecc9ebc9ef367ab9c9):


```
=== RUN   TestWindowFunctionsAgainstProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWindowFunctionsAgainstProcessor377914869
    test_log_scope.go:74: use -show-logs to present logs inline
    panic.go:965: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWindowFunctionsAgainstProcessor377914869
    panic.go:965: unsupported comparison: date to timestamp
        goroutine 25338 [running]:
        runtime/debug.Stack(0xc0003ea860, 0x3deb340, 0xc000163020)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0x27e
        panic(0x3deb340, 0xc000163020)
        	/usr/local/go/src/runtime/panic.go:965 +0x1b9
        github.com/cockroachdb/cockroach/pkg/sql/sem/tree.compareTimestamps(0xc0010786c0, 0x56140e8, 0xc000a6bc10, 0x5614958, 0xc001126348, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:2339 +0x408
        github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DDate).Compare(0xc000a6bc10, 0xc0010786c0, 0x5614958, 0xc001126348, 0x56140e8)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1838 +0xc6
        github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*WindowFrameRun).FrameStartIdx.func4(0xb, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/window_funcs.go:187 +0xf5
        sort.Search(0xc, 0xc0003eac48, 0xc000072108)
        	/usr/local/go/src/sort/search.go:66 +0x58
        github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*WindowFrameRun).FrameStartIdx(0xc0010a8690, 0x54e7928, 0xc000072108, 0xc0010786c0, 0x0, 0x1, 0x54cb850)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/window_funcs.go:178 +0x627
        github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.lastValueWindow.Compute(0x54e7928, 0xc000072108, 0xc0010786c0, 0xc0010a8690, 0x0, 0x0, 0x9, 0xc0010fc310)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/window_builtins.go:772 +0x4d
        github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*windower).processPartition(0xc0010c8a80, 0x54e7928, 0xc000072108, 0xc0010786c0, 0xc0010d2700, 0x0, 0x0, 0x8)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/windower.go:564 +0x306
        github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*windower).computeWindowFunctions(0xc0010c8a80, 0x54e7928, 0xc000072108, 0xc0010786c0, 0x0, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/windower.go:679 +0xaed
        github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*windower).emitRow(0xc0010c8a80, 0x2, 0x0, 0x0, 0x0, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/windower.go:313 +0x92
        github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*windower).Next(0xc0010c8a80, 0x47ee6a0, 0xc001159a20, 0xc0010cbbc0, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/windower.go:223 +0x59
        github.com/cockroachdb/cockroach/pkg/sql/distsql.verifyColOperator(0xc00142af00, 0x1, 0x0, 0x0, 0x0, 0xc0010b9a40, 0x1, 0x1, 0xc0010b9a58, 0x1, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/columnar_utils_test.go:191 +0xfff
        github.com/cockroachdb/cockroach/pkg/sql/distsql.TestWindowFunctionsAgainstProcessor(0xc00142af00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/columnar_operators_test.go:1185 +0x998
        testing.tRunner(0xc00142af00, 0x47ee818)
        	/usr/local/go/src/testing/testing.go:1193 +0xef
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1238 +0x2b3
--- FAIL: TestWindowFunctionsAgainstProcessor (1.20s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestWindowFunctionsAgainstProcessor PKG=./pkg/sql/distsql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries yuzefovich
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWindowFunctionsAgainstProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/window
+++ b/pkg/sql/logictest/testdata/logic_test/window
@@ -3947,3 +3947,17 @@ FROM (VALUES (NULL::INT), (NULL::INT), (1)) v(x);
 1
 1
 NULL
+
+# Regression test for panic during comparison for RANGE mode with
+# OFFSET PRECEDING or OFFSET FOLLOWING (#67975).
+statement ok
+DROP TABLE IF EXISTS t;
+CREATE TABLE t (x DATE);
+INSERT INTO t VALUES ('5874897-01-01'::DATE), ('1999-01-08'::DATE);
+SET vectorize=off;
+
+query error timestamp "5874897-01-01T00:00:00Z" exceeds supported timestamp bounds
+SELECT first_value(x) OVER (ORDER BY x RANGE BETWEEN CURRENT ROW AND '0 YEAR'::INTERVAL FOLLOWING) FROM t;
+
+statement ok
+RESET vectorize;

--- a/pkg/sql/sem/tree/datum.go
+++ b/pkg/sql/sem/tree/datum.go
@@ -2262,31 +2262,31 @@ func (*DTimestamp) ResolvedType() *types.T {
 
 // timeFromDatumForComparison gets the time from a datum object to use
 // strictly for comparison usage.
-func timeFromDatumForComparison(ctx *EvalContext, d Datum) (time.Time, bool) {
+func timeFromDatumForComparison(ctx *EvalContext, d Datum) (time.Time, error) {
 	d = UnwrapDatum(ctx, d)
 	switch t := d.(type) {
 	case *DDate:
 		ts, err := MakeDTimestampTZFromDate(ctx.GetLocation(), t)
 		if err != nil {
-			return time.Time{}, false
+			return time.Time{}, err
 		}
-		return ts.Time, true
+		return ts.Time, nil
 	case *DTimestampTZ:
-		return t.Time, true
+		return t.Time, nil
 	case *DTimestamp:
 		// Normalize to the timezone of the context.
 		_, zoneOffset := t.Time.In(ctx.GetLocation()).Zone()
 		ts := t.Time.In(ctx.GetLocation()).Add(-time.Duration(zoneOffset) * time.Second)
-		return ts, true
+		return ts, nil
 	case *DTime:
 		// Normalize to the timezone of the context.
 		toTime := timeofday.TimeOfDay(*t).ToTime()
 		_, zoneOffsetSecs := toTime.In(ctx.GetLocation()).Zone()
-		return toTime.In(ctx.GetLocation()).Add(-time.Duration(zoneOffsetSecs) * time.Second), true
+		return toTime.In(ctx.GetLocation()).Add(-time.Duration(zoneOffsetSecs) * time.Second), nil
 	case *DTimeTZ:
-		return t.ToTime(), true
+		return t.ToTime(), nil
 	default:
-		return time.Time{}, false
+		return time.Time{}, errors.AssertionFailedf("unexpected type: %v", t.ResolvedType())
 	}
 }
 
@@ -2333,9 +2333,9 @@ func compareTimestamps(ctx *EvalContext, l Datum, r Datum) int {
 		// values to get the desired result for comparison.
 		return int(leftInf - rightInf)
 	}
-	lTime, lOk := timeFromDatumForComparison(ctx, l)
-	rTime, rOk := timeFromDatumForComparison(ctx, r)
-	if !lOk || !rOk {
+	lTime, lErr := timeFromDatumForComparison(ctx, l)
+	rTime, rErr := timeFromDatumForComparison(ctx, r)
+	if lErr != nil || rErr != nil {
 		panic(makeUnsupportedComparisonMessage(l, r))
 	}
 	if lTime.Before(rTime) {

--- a/pkg/sql/sem/tree/window_funcs.go
+++ b/pkg/sql/sem/tree/window_funcs.go
@@ -13,6 +13,7 @@ package tree
 import (
 	"context"
 	"sort"
+	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
@@ -133,7 +134,12 @@ func (wfr *WindowFrameRun) FrameStartIdx(ctx context.Context, evalCtx *EvalConte
 						wfr.err = err
 						return false
 					}
-					return valueAt.Compare(evalCtx, value) <= 0
+					cmp, err := compareForWindow(evalCtx, valueAt, value)
+					if err != nil {
+						wfr.err = err
+						return false
+					}
+					return cmp <= 0
 				}), wfr.err
 			}
 			// We use binary search on [0, wfr.RowIdx) interval to find the first row
@@ -148,7 +154,12 @@ func (wfr *WindowFrameRun) FrameStartIdx(ctx context.Context, evalCtx *EvalConte
 					wfr.err = err
 					return false
 				}
-				return valueAt.Compare(evalCtx, value) >= 0
+				cmp, err := compareForWindow(evalCtx, valueAt, value)
+				if err != nil {
+					wfr.err = err
+					return false
+				}
+				return cmp >= 0
 			}), wfr.err
 		case CurrentRow:
 			// Spec: in RANGE mode CURRENT ROW means that the frame starts with the current row's first peer.
@@ -170,7 +181,12 @@ func (wfr *WindowFrameRun) FrameStartIdx(ctx context.Context, evalCtx *EvalConte
 						wfr.err = err
 						return false
 					}
-					return valueAt.Compare(evalCtx, value) <= 0
+					cmp, err := compareForWindow(evalCtx, valueAt, value)
+					if err != nil {
+						wfr.err = err
+						return false
+					}
+					return cmp <= 0
 				}), wfr.err
 			}
 			// We use binary search on [0, wfr.PartitionSize()) interval to find the
@@ -184,7 +200,12 @@ func (wfr *WindowFrameRun) FrameStartIdx(ctx context.Context, evalCtx *EvalConte
 					wfr.err = err
 					return false
 				}
-				return valueAt.Compare(evalCtx, value) >= 0
+				cmp, err := compareForWindow(evalCtx, valueAt, value)
+				if err != nil {
+					wfr.err = err
+					return false
+				}
+				return cmp >= 0
 			}), wfr.err
 		default:
 			return 0, errors.AssertionFailedf(
@@ -303,7 +324,12 @@ func (wfr *WindowFrameRun) FrameEndIdx(ctx context.Context, evalCtx *EvalContext
 						wfr.err = err
 						return false
 					}
-					return valueAt.Compare(evalCtx, value) < 0
+					cmp, err := compareForWindow(evalCtx, valueAt, value)
+					if err != nil {
+						wfr.err = err
+						return false
+					}
+					return cmp < 0
 				}), wfr.err
 			}
 			// We use binary search on [0, wfr.PartitionSize()) interval to find
@@ -320,7 +346,12 @@ func (wfr *WindowFrameRun) FrameEndIdx(ctx context.Context, evalCtx *EvalContext
 					wfr.err = err
 					return false
 				}
-				return valueAt.Compare(evalCtx, value) > 0
+				cmp, err := compareForWindow(evalCtx, valueAt, value)
+				if err != nil {
+					wfr.err = err
+					return false
+				}
+				return cmp > 0
 			}), wfr.err
 		case CurrentRow:
 			// Spec: in RANGE mode CURRENT ROW means that the frame end with the current row's last peer.
@@ -342,7 +373,12 @@ func (wfr *WindowFrameRun) FrameEndIdx(ctx context.Context, evalCtx *EvalContext
 						wfr.err = err
 						return false
 					}
-					return valueAt.Compare(evalCtx, value) < 0
+					cmp, err := compareForWindow(evalCtx, valueAt, value)
+					if err != nil {
+						wfr.err = err
+						return false
+					}
+					return cmp < 0
 				}), wfr.err
 			}
 			// We use binary search on [0, wfr.PartitionSize()) interval to find
@@ -356,7 +392,12 @@ func (wfr *WindowFrameRun) FrameEndIdx(ctx context.Context, evalCtx *EvalContext
 					wfr.err = err
 					return false
 				}
-				return valueAt.Compare(evalCtx, value) > 0
+				cmp, err := compareForWindow(evalCtx, valueAt, value)
+				if err != nil {
+					wfr.err = err
+					return false
+				}
+				return cmp > 0
 			}), wfr.err
 		case UnboundedFollowing:
 			return wfr.unboundedFollowing(), nil
@@ -633,6 +674,25 @@ func (wfr *WindowFrameRun) IsRowSkipped(ctx context.Context, idx int) (bool, err
 	}
 	// If a row is excluded from the window frame, it is skipped.
 	return wfr.isRowExcluded(idx)
+}
+
+// compareForWindow wraps the Datum Compare method so that casts can be
+// performed up front. This allows us to return an expected error in the event
+// of an invalid comparison, rather than panicking.
+func compareForWindow(evalCtx *EvalContext, left, right Datum) (int, error) {
+	if types.IsDateTimeType(left.ResolvedType()) && left.ResolvedType() != types.Interval {
+		// Datetime values (other than Intervals) are converted to timestamps for
+		// comparison. Note that the right side never needs to be casted.
+		ts, err := timeFromDatumForComparison(evalCtx, left)
+		if err != nil {
+			return 0, err
+		}
+		left, err = MakeDTimestampTZ(ts, time.Microsecond)
+		if err != nil {
+			return 0, err
+		}
+	}
+	return left.Compare(evalCtx, right), nil
 }
 
 // WindowFunc performs a computation on each row using data from a provided *WindowFrameRun.

